game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java

```diff
@@ -28,37 +29,36 @@
  */
 public class ExecutionStack implements Serializable {
   private static final long serialVersionUID = -8675285470515074530L;
-  private IExecutable m_current;
-  @SuppressWarnings("JdkObsolete") // change to Deque/ArrayDeque upon next incompatible release
-  private final Stack<IExecutable> m_stack = new Stack<>();
+  private boolean isExecuting;
+  private final Deque<IExecutable> deque = new ArrayDeque<>();
 
   void execute(final IDelegateBridge bridge) {
-    // we were interrupted before, resume where we left off
-    if (m_current != null) {
-      m_stack.push(m_current);
+    isExecuting = true;
+    while (!deque.isEmpty()) {
+      final IExecutable executable = deque.pop();
+      try {
+        executable.execute(this, bridge);
+      } catch (final RuntimeException e) {
+        deque.push(executable);
+        throw e;
+      }
     }
-    while (!m_stack.isEmpty()) {
-      m_current = m_stack.pop();
-      m_current.execute(this, bridge);
-    }
-    m_current = null;
+    isExecuting = false;
   }
 
   void push(final Collection<IExecutable> executables) {
-    for (final IExecutable ex : executables) {
-      push(ex);
-    }
+    executables.forEach(deque::push);
   }
 
   public void push(final IExecutable executable) {
-    m_stack.push(executable);
+    deque.push(executable);
   }
 
   boolean isExecuting() {
-    return m_current != null;
+    return isExecuting;
   }
 
   public boolean isEmpty() {
-    return m_stack.isEmpty();
+    return deque.isEmpty();
```

Comment:
This could cause different behavior if its called before execute() is or after execution is interrupted and the current one is added back. Did you double check those scenarios?

Reply:
Good point, I did not.
I'll try figuring out if there are cases when an exception is caught, where - as a response to that - executables are added to the stack, later this evening.